docs/RationaleSeq.md

```diff
@@ -0,0 +1,100 @@
+  clock. Transparent the entire time the clock is high. Generally referred to
+  as a "latch". Examples: "gated SR latch", "D latch".
+  - Edge-triggered **flip-flop:** An edge-sensitive memory element. Captures
+  the input value on one or both clock edges. Variants: posedge FF, negedge
+  FF, "edge-sensitive" FF (captures the input value on both edges),
+  resettable FF.
+- Abstract models:
+  - **Register:** A synchronous, resettable memory element. Can be
+  implemented using any of the above "circuit level" elements.
+
+## The register operation
+
+The `seq.reg` op models an abstract notion of a "register", independent of
+its implementation (e.g. latch, D flip-flop). It is intended to be 'lowered'
+to a specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic. Our
+intention is to allow analysis and optimization of sequential logic without
+having to reason about timing-dependent behavior. This makes it somewhat
+distinct from the SystemVerilog register wherein one has to worry about reset
+behavior and implementation (latch vs. flip-flop).
+
+`Reg` has four operands:
+
+- **input**: The value to be captured at 'clock'. Generally called 'q'.
+Accepts any type, results in the same type.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A value which the state is set to upon reset. Required iff
+'reset' is present.
+
+```mlir
+%d = seq.reg %input, %clk [, %reset [, %resetValue] ] : $type(input)
+```
+
+### Rationale
+
+Several design decisions were made in defining this op. Mostly, they were
+made to simplify it while still providing the common case. Providing support
+for all flavors of registers is an anti-goal of this op. If an omitted
+feature is needed, it can be added or (if not common or precludes
+optimization) another op could be added.
+
+- Logical features:
+  - Inclusion of optional 'reset' signal: This operand makes lowering to an
+  efficient implementation of reset easier. Omission of it would require some
+  (potentially complex) analysis to find the reset mux if required (which it
+  usually is to ensure efficient synthesis).
+  - Inclusion of 'resetValue': if we have a 'reset' signal, we need to
+  include a value.
+  - Omission of 'enable': enables are easily modeled via a multiplexer on the
+  input with one of the mux inputs as the output of the register. This
+  property makes 'enables' easier to detect than reset in the common case.
+
+- Timing / clocking:
+  - Omission of 'negedge' event on 'clock': this is easily modeled by
+  inverting the clock value.
+  - Omission of 'dual edge' event on 'clock': this is not expected to be
+  terribly common.
+  - Omission of edge conditions on 'reset': The reset style shouldn't affect
+  logical correctness. It should, therefore, be determined by a lowering
+  pass.
```

Comment:
We do need to know async/sync for codegen, so is this op encoded or type encoded?

Reply:
IMO, this is a style thing. For any given CIRCT design, one _generally_ wants the same reset type. If we find that's not the case we can talk about modeling reset types.

Reply:
i don't think that's a good assumption to make.

Reply:
Why? HLS workloads certainly won't care. I can't find a place in the FIRRTL IR where one can specific async vs sync reset.

Reply:
In FIRRTL IR it is inferred from the reset signal itself (which is reset or async). I'm not saying that's how it should be done, but it certainly "knows" at the time of the emission. If that's how you want to do it also I suggest adding that to the rationale doc.

personally I might want some justification why we want to represent sync reset at all, since in theory that could always be pulled back into the logic. Not sure sync reset flops are a real hardware thing.

Reply:
There is some documentation on how this works here: https://www.chisel-lang.org/chisel3/docs/wiki-deprecated/reset.html. tl;dr: Three types of reset (which are represented in FIRRTL IR):

Reply:
> In FIRRTL IR it is inferred from the reset signal itself (which is reset or async typed). I'm not saying that's how it should be done, but it certainly "knows" at the time of the emission. If that's how you want to do it also I suggest adding that to the rationale doc.

I'm not sure if that's the way we want to go or not -- we have yet to discuss type based vs attribute based for clock and reset styles. Also bear in mind that this will **not** be directly emitted by ExportVerilog. It will be lowered into another dialect first by any lowering pass which can decide on the actual implementation.

> NOTE: by "FIRRTL IR" I am referring to the scala implementation, which I realized might not be what you were referring to.

I was just quickly skimming the FIRRTL dialect code. By FIRRTL IR, I was referring to the FIRRTL dialect. @lattner / @darthscsi Are reset types currently supported?

> personally I might want some justification why we want to represent sync reset at all, since in theory that could always be pulled back into the logic. Not sure sync reset flops are a real hardware thing.

Not to be flippant, but nothing besides doped sand is an actual hardware thing. After that, it's all just levels of abstraction. This is not intended to be directly emitted to hardware. It's intended to support that sweet spot at which relatively high-level modeling is possible (e.g. LLHD should be able to compile fast and accurate simulators) and lowering to an actual implementation is possible. It should make the simple, common case easy to reason about.

Reply:
> I do note that this is one solution that we fell into. However, it's perilous to assume that all registers in a compilation unit have the same reset type.

That's part of my concern of modeling reset types too soon. We need to carefully consider the abstraction we provide at this level, which I don't think is absolutely necessary to start with.

Reply:
I'm pretty sure every single design needs async reset, even if the rest of the design follows sync reset. Synchronizers, clock domain crossing FIFOs, etc, so if you model sync reset as reset at all, then you need to distinguish. If the goal is to keep it as simple as possible, then I think @mwachs5's suggestion of just not modeling sync reset as reset is the best option.

Reply:
@fabianschuiki Are there any issues with async reset modeling in LLHD? Would it cause any performance issues with the generated simulators? (esp. for what I continue to think is the common case of either don't care or sync reset.)

Reply:
> I'm pretty sure every single design* needs async reset, even if the rest of the design follows a sync reset discipline. Synchronizers, clock domain crossing FIFOs, etc, so if you model sync reset as reset at all, then you need to distinguish.

I'm not arguing that this op could be use to model every use of SystemVerilog registers. Rather, I intended it to model  completely synchronous "logic support" registers. E.g. FSM regs, pipeline regs. In these cases, you generally don't care about the reset behavior from a correctness standpoint -- adding specification of the reset behavior just introduces unnecessary complexity for this use case. It's like the Chisel notion of "Reset-Agnostic Code", which I suspect is the most common case in computational cores.

CDC registers are somewhat different and I think it makes sense to model them differently at this level. They often need to be coded in very particular ways and/or accompanied by a timing constraints file, which makes them different from a purely mechanical standpoint. They also should be (I suspect) optimized differently. In the general case, optimizations shouldn't touch CDC regs as pretty much any optimization would break them.